packages/fxa-shared/metrics/amplitude-event.1.schema.json

```diff
@@ -197,7 +235,72 @@
         }
       },
       "then": {
-        "required": ["language"]
+        "required": ["language"],
```

Comment:
I know in a previous patch, we'd added this `allOf` block, but all of these `if` checks get checked for every event this way, which is not performant considering all FxA events get checked too each time.

Also, it's hard to reason about what applies to which event or set of events with everything at the same block level in the schema.

Can I suggest a different approach?

Instead of:
```
allOf: [{
  {
    if: ...,
    then: ...
  },
  {
    if: ...,
    then: ...
  },
  ...
}]
```
How about:
```md
- if pattern matches `fxa_pay_setup - | fxa_pay_subscription_change - `
	- if pattern matches `fxa_pay_setup -`
		- if pattern matches `ends with view`
		- if pattern matches `ends with engage`
		- if pattern matches `ends with submit`
		- if pattern matches `ends with fail`
		- if pattern matches `ends with success`
	- Else (pattern matches `fxa_pay_subscription_change -`)
		- if pattern matches `ends with view`
		- if pattern matches `ends with engage`
		- if pattern matches `ends with submit`
		- if pattern matches `ends with fail`
		- if pattern matches `ends with success`
```

This is a similar suggestion to the unit test comment suggestion of nested `describe` blocks I made in my first review pass.

Since it's kind of a pain to work with JSON schemas, I already have a diff you can apply and try out. With nested `if` blocks, any `event_type` that doesn't match that first condition won't be checked any further. Note: I put a bunch of `"required": ["plan_id"]` lines in some places just to get `ajv` to compile the schema successfully, since you can't have an empty `"required": []`. In those cases where an event doesn't have any required properties other than what's already covered above it, I guess we can omit it or have some stub if it can compile successfully to acknowledge it exists?

```diff
diff --git a/packages/fxa-shared/metrics/amplitude-event.1.schema.json b/packages/fxa-shared/metrics/amplitude-event.1.schema.json
index bb0d393ff4..ff650f0497 100644
--- a/packages/fxa-shared/metrics/amplitude-event.1.schema.json
+++ b/packages/fxa-shared/metrics/amplitude-event.1.schema.json
@@ -56,7 +56,7 @@
       "type": "string",
       "pattern": "^[a-z]{1,}(-[a-zA-Z0-9]{1,})*$"
     },
-    "countryCode": {
+    "country_code": {
       "description": "The 2 character country ISO code",
       "type": "string",
       "pattern": "^[A-Z]{2}$"
@@ -117,6 +117,10 @@
           "type": "string",
           "enum": ["with-account", "without-account"]
         },
+        "error_id": {
+          "description": "TODO add description",
+          "type": "string"
+        },
         "previous_plan_id": {
           "description": "The Stripe price_* or plan_* ID for the subscription the user currently has prior to the change",
           "type": "string"
@@ -222,86 +226,244 @@
       "required": ["device_id"]
     }
   ],
-  "allOf": [
-    {
-      "if": {
-        "$comment": "For event_type's matching the pattern below, make following fields required",
-        "required": ["event_type"],
-        "properties": {
-          "event_type": {
-            "type": "string",
-            "pattern": "^fxa_pay_setup -|fxa_pay_subscription_change -"
-          }
-        }
+  "if": {
+    "$comment": "For all 'fxa_pay_setup - *' and 'fxa_pay_subscription_change - *' events (i.e. all payments server conversion funnel events)",
+    "required": ["event_type"],
+    "properties": {
+      "event_type": {
+        "type": "string",
+        "pattern": "^fxa_pay_setup -|fxa_pay_subscription_change -"
+      }
+    }
+  },
+  "then": {
+    "required": ["language"],
+    "properties": {
+      "user_properties": {
+        "required": ["flow_id"]
       },
-      "then": {
-        "required": ["language"],
-        "properties": {
-          "user_properties": {
-            "required": ["flow_id"]
-          },
-          "event_properties": {
-            "required": ["plan_id", "product_id"]
-          }
-        }
+      "event_properties": {
+        "required": ["plan_id", "product_id", "checkout_type"]
       }
     },
-    {
-      "if": {
-        "$comment": "For event_type's matching the pattern below, make following fields required",
-        "required": ["event_type"],
-        "properties": {
-          "event_type": {
-            "type": "string",
-            "pattern": "^fxa_pay_setup -"
-          }
-        }
-      },
-      "then": {
-        "properties": {
-          "event_properties": {
-            "required": ["checkout_type"]
-          }
+    "if": {
+      "$comment": "For all fxa_pay_setup (new subscription) funnel events",
+      "required": ["event_type"],
+      "properties": {
+        "event_type": {
+          "type": "string",
+          "pattern": "^fxa_pay_setup -"
         }
       }
     },
-    {
-      "if": {
-        "$comment": "For event_type's matching the pattern below, make following fields required",
-        "required": ["event_type"],
-        "properties": {
-          "event_type": {
-            "type": "string",
-            "pattern": "^fxa_pay_setup - submit|fxa_pay_setup - success|fxa_pay_setup - fail"
-          }
+    "then": {
+      "properties": {
+        "event_properties": {
+          "required": ["checkout_type"]
         }
       },
-      "then": {
-        "properties": {
-          "event_properties": {
-            "required": ["payment_provider"]
+      "allOf": [
+        {
+          "if": {
+            "$comment": "For all 'fxa_pay_setup - view' subevents",
+            "required": ["event_type"],
+            "properties": {
+              "event_type": {
+                "type": "string",
+                "pattern": "$view"
+              }
+            }
+          },
+          "then": {
+            "properties": {
+              "event_properties": {
+                "required": ["plan_id"]
+              }
+            }
+          }
+        },
+        {
+          "if": {
+            "$comment": "For all 'fxa_pay_setup - engage' subevents",
+            "required": ["event_type"],
+            "properties": {
+              "event_type": {
+                "type": "string",
+                "pattern": "$engage"
+              }
+            }
+          },
+          "then": {
+            "properties": {
+              "event_properties": {
+                "required": ["plan_id"]
+              }
+            }
+          }
+        },
+        {
+          "if": {
+            "$comment": "For all 'fxa_pay_setup - submit' subevents",
+            "required": ["event_type"],
+            "properties": {
+              "event_type": {
+                "type": "string",
+                "pattern": "$submit"
+              }
+            }
+          },
+          "then": {
+            "properties": {
+              "event_properties": {
+                "required": ["plan_id"]
+              }
+            }
+          }
+        },
+        {
+          "if": {
+            "$comment": "For all 'fxa_pay_setup - success' subevents",
+            "required": ["event_type"],
+            "properties": {
+              "event_type": {
+                "type": "string",
+                "pattern": "$success"
+              }
+            }
+          },
+          "then": {
+            "properties": {
+              "event_properties": {
+                "required": ["plan_id"]
+              }
+            }
+          }
+        },
+        {
+          "if": {
+            "$comment": "For all 'fxa_pay_setup - fail' subevents",
+            "required": ["event_type"],
+            "properties": {
+              "event_type": {
+                "type": "string",
+                "pattern": "$fail"
+              }
+            }
+          },
+          "then": {
+            "properties": {
+              "event_properties": {
+                "required": ["error_id"]
+              }
+            }
           }
         }
-      }
+      ]
     },
-    {
-      "if": {
-        "$comment": "For event_type's matching the pattern below, make following fields required",
-        "required": ["event_type"],
-        "properties": {
-          "event_type": {
-            "type": "string",
-            "pattern": "^fxa_pay_subscription_change -"
-          }
+    "else": {
+      "$comment": "For all fxa_pay_subscription_change (update subscription) funnel events",
+      "properties": {
+        "event_properties": {
+          "required": ["previous_plan_id", "previous_product_id"]
         }
       },
-      "then": {
-        "properties": {
-          "event_properties": {
-            "required": ["previous_plan_id", "previous_product_id", "subscription_id", "payment_provider"]
+      "allOf": [
+        {
+          "if": {
+            "$comment": "For all 'fxa_pay_subscription_change - view' subevents",
+            "required": ["event_type"],
+            "properties": {
+              "event_type": {
+                "type": "string",
+                "pattern": "$view"
+              }
+            }
+          },
+          "then": {
+            "properties": {
+              "event_properties": {
+                "required": ["plan_id"]
+              }
+            }
+          }
+        },
+        {
+          "if": {
+            "$comment": "For all 'fxa_pay_subscription_change - engage' subevents",
+            "required": ["event_type"],
+            "properties": {
+              "event_type": {
+                "type": "string",
+                "pattern": "$engage"
+              }
+            }
+          },
+          "then": {
+            "properties": {
+              "event_properties": {
+                "required": ["plan_id"]
+              }
+            }
+          }
+        },
+        {
+          "if": {
+            "$comment": "For all 'fxa_pay_subscription_change - submit' subevents",
+            "required": ["event_type"],
+            "properties": {
+              "event_type": {
+                "type": "string",
+                "pattern": "$submit"
+              }
+            }
+          },
+          "then": {
+            "properties": {
+              "event_properties": {
+                "required": ["plan_id"]
+              }
+            }
+          }
+        },
+        {
+          "if": {
+            "$comment": "For all 'fxa_pay_subscription_change - success' subevents",
+            "required": ["event_type"],
+            "properties": {
+              "event_type": {
+                "type": "string",
+                "pattern": "$success"
+              }
+            }
+          },
+          "then": {
+            "properties": {
+              "event_properties": {
+                "required": ["plan_id"]
+              }
+            }
+          }
+        },
+        {
+          "if": {
+            "$comment": "For all 'fxa_pay_subscription_change - fail' subevents",
+            "required": ["event_type"],
+            "properties": {
+              "event_type": {
+                "type": "string",
+                "pattern": "$fail"
+              }
+            }
+          },
+          "then": {
+            "properties": {
+              "event_properties": {
+                "required": ["plan_id"]
+              }
+            }
           }
         }
-      }
+      ]
     }
-  ]
+  }
 }

```

Reply:
My initial thoughts are that we might be prematurely optimizing the code for performance concerns at the expense of readability/maintainability. The original implementation was written to avoid nested conditionals. While nested conditionals will save some computations for some events, it will add checks for others. v8's regular expression system is pretty fast and ajv compiles schemas so that they are already optimized at runtime.

I'm open to your thoughts here, perhaps I can schedule a call?